### PR TITLE
allow for extensions of valkyrie indexer's to_solr that depends on resource custom metadata

### LINKED ISSUE #4464: indexer shared specs are not usable outside hyrax when they depend on resource custom fields
### Descriptive summary

All the indexer shared specs create the resource for the spec to use in its testing.  They set values in the resource which are tested for in the shared spec.

If an indexer overrides #to_solr and expects to be able to use a property field from the class of resource that it indexes, the call to #to_solr fail because the resource (e.g. Hyrax::Resource, Hyrax::Work, Hyrax::PcdmCollection) does not have that property defined.  In some cases, the property is required and should fail if missing.  Additionally, sometimes the value of the property is more than a simple value.  If the #to_solr customization expects that property to have a value and uses a part of that property (e.g. my_property.first), it will fail if an appropriate value is not set.

### Rationale

The shared spec cannot know customized requirements of the indexer.  As such, it cannot create a resource that will allow the #to_solr method to complete without error.

### Expected behavior

For a resource indexer with customizations to #to_solr that use a property defined in the resource, the shared spec `it_behaves_like 'a Hyrax::Resource indexer'` should pass.

### Actual behavior

For a resource indexer with customizations to #to_solr that use a property defined in the resource, the shared spec `it_behaves_like 'a Hyrax::Resource indexer'` fails with error similar to...

```
NoMethodError:
       undefined method `isbn' for #<Hyrax::Resource:0x00007f860abe0180>
```

### Steps to reproduce the behavior

See [when extending with custom metadata](https://github.com/samvera/hyrax/blob/9a8f725302934a6b95e52451c3d8598e0a003b85/spec/indexers/hyrax/valkyrie_work_indexer_spec.rb#L24) test in PR #4463.

NOTE: This effects all shared indexer specs and the solution needs to verify that all shared specs allow for customizations to #to_solr.

### PR DESCRIPTION
**Pull request by elrayle**
Fixes #4464

### Description

Allow shared spec testing of indexers with customizations to #to_solr that depend on properties defined in the resource class that the indexer indexes.

### Resolution

Update shared indexers spec to have a valid resource passed in to the shared spec.  The shared spec sets values it depends on in a before block.  

@samvera/hyrax-code-reviewers

### CODE REVIEW

File: lib/hyrax/specs/shared_specs/indexers.rb [hunk 1]
```diff
@@ -3,7 +3,8 @@
 RSpec.shared_examples 'a Hyrax::Resource indexer' do
   subject(:indexer) { indexer_class.new(resource: resource) }
   let(:ids)         { ['id1', 'id2'] }
-  let(:resource)    { Hyrax::Resource.new(alternate_ids: ids) }
+  let(:the_resource_class) { defined?(resource_class) ? resource_class : Hyrax::Work }
```
Comment (elrayle) [inline lib/hyrax/specs/shared_specs/indexers.rb:6]: Setting resource_class to Hyrax::Work when it isn't defined allows for backward compatibility.

File: spec/indexers/hyrax/valkyrie_collection_indexer_spec.rb [hunk 2]
```diff
@@ -3,7 +3,12 @@
 require 'hyrax/specs/shared_specs'
 
 RSpec.describe Hyrax::ValkyrieCollectionIndexer do
-  subject(:indexer) { described_class.new(resource: resource) }
```
Comment (elrayle) [inline spec/indexers/hyrax/valkyrie_collection_indexer_spec.rb:6]: The subject is not used anywhere, so it was removed.

File: spec/indexers/hyrax/valkyrie_collection_indexer_spec.rb [hunk 3]
```diff
@@ -3,7 +3,12 @@
...
+  let(:resource_class) do
+    Class.new(Hyrax::Resource) do
+      # include ::Hyrax::CollectionBehavior # TODO: How is this represented in the Valkyrie::Resource version of a Collection?
```
Comment (elrayle) [inline spec/indexers/hyrax/valkyrie_collection_indexer_spec.rb:8]: I could not find any example of a defined or generated Collection that is a Valkyrie::Resource.  The AF version includes CollectionBehavior.  I commented it out and left it here as a placeholder for any behaviors that are expected/required for a Valkyrie version of a Hyrax Collection.  I'd be interested in what others think should be included in this class.  NOTE: Prior to this PR, the Collection indexer was being tested with a Hyrax::Work as the resource passed to the indexer.

File: spec/indexers/hyrax/valkyrie_work_indexer_spec.rb [hunk 4]
```diff
@@ -3,8 +3,7 @@
 require 'hyrax/specs/shared_specs'
 
 RSpec.describe Hyrax::ValkyrieWorkIndexer do
-  let(:indexer) { described_class.new(resource: resource) }
-  let(:resource) { FactoryBot.valkyrie_create(:hyrax_work) }
```
Comment (elrayle) [inline spec/indexers/hyrax/valkyrie_work_indexer_spec.rb:7]: indexer and resource were not used by this test, so they were removed.

File: spec/indexers/hyrax/valkyrie_work_indexer_spec.rb [hunk 5]
```diff
@@ -21,4 +20,11 @@
 
     it_behaves_like 'a Basic metadata indexer'
   end
+
+  context 'when extending with custom metadata' do
```
Comment (elrayle) [inline spec/indexers/hyrax/valkyrie_work_indexer_spec.rb:24]: This context is the spec that is addressing the Actual Behavior vs. Expected Behavior described in the PR comments.

File: spec/support/book_resource.rb [hunk 6]
```diff
@@ -21,6 +21,14 @@ class Book < ActiveFedora::Base
       property :publisher, predicate: ::RDF::URI('http://example.com/ns/publisher')
       property :title,     predicate: ::RDF::URI("http://example.com/ns/title")
     end
+
+    class BookResourceIndexer < Hyrax::ValkyrieWorkIndexer
+      def to_solr
```
Comment (elrayle) [inline spec/support/book_resource.rb:26]: This defines an indexer that uses custom metadata from the resource.

File: lib/hyrax/specs/shared_specs/indexers.rb (hunk 1 above)
Comment (elrayle) [reply lib/hyrax/specs/shared_specs/indexers.rb:6]: Passing in the resource_class allows #to_solr to successfully create the solr_doc when the indexer is expecting the resource to have custom metadata defined.

File: spec/indexers/hyrax/valkyrie_collection_indexer_spec.rb (hunk 3 above)
Comment (elrayle) [reply spec/indexers/hyrax/valkyrie_collection_indexer_spec.rb:8]: Found it, this should be Hyrax::PcdmCollection

File: lib/hyrax/specs/shared_specs/indexers.rb [hunk 7]
```diff
@@ -1,11 +1,12 @@
 # frozen_string_literal: true
 
 RSpec.shared_examples 'a Hyrax::Resource indexer' do
-  subject(:indexer) { indexer_class.new(resource: resource) }
-  let(:ids)         { ['id1', 'id2'] }
-  let(:resource)    { Hyrax::Resource.new(alternate_ids: ids) }
+  subject(:indexer)  { indexer_class.new(resource: the_resource) }
+  let(:ids)          { ['id1', 'id2'] }
+  let(:the_resource) { defined?(resource) ? resource : Hyrax::Resource.new }
```
Comment (elrayle) [inline lib/hyrax/specs/shared_specs/indexers.rb:6]: This is providing backward compatibility.  If the caller does not create a resource, then one is created in the shared spec.  If backward compatibility is not required, this setup would become...

```
   subject(:indexer)  { indexer_class.new(resource: resource) }
   let(:ids)          { ['id1', 'id2'] }
```

NOTE: All references to `the_resource` are providing backward compatibility for each shared spec.  Similar changes could occur if `resource` was required to be defined by the caller.

File: lib/hyrax/specs/shared_specs/indexers.rb [hunk 8]
```diff
@@ -1,11 +1,12 @@
...
 
   describe '#to_solr' do
+    before { the_resource.alternate_ids = ids }
```
Comment (elrayle) [inline lib/hyrax/specs/shared_specs/indexers.rb:9]: Makes sure the alternate_ids are set whether the resource is created in the shared spec or by the caller.

NOTE: Similar use of before blocks in other shared specs to be sure the resource has the values set that are being tested.

File: spec/indexers/hyrax/valkyrie_work_indexer_spec.rb [hunk 9]
```diff
@@ -21,4 +20,13 @@
...
+    let(:resource) { Hyrax::Test::BookResource.new(isbn: '978-3-16-148410-0', publisher: 'The Publisher', title: 'A title') }
+    let(:indexer_class) { Hyrax::Test::BookResourceIndexer }
+
+    it_behaves_like 'a Hyrax::Resource indexer'
+    it_behaves_like 'a permission indexer'
```
Comment (elrayle) [inline spec/indexers/hyrax/valkyrie_work_indexer_spec.rb:29]: permission indexer tests currently failing with...
```
NoMethodError:
     #   undefined method `internal_resource' for nil:NilClass
```

File: spec/indexers/hyrax/valkyrie_work_indexer_spec.rb [hunk 10]
```diff
@@ -21,4 +20,13 @@
...
+    it_behaves_like 'a visibility indexer'
+  end
```
Comment (elrayle) [inline spec/indexers/hyrax/valkyrie_work_indexer_spec.rb:31]: permission indexer tests currently failing with...
```
NoMethodError:
     #   undefined method `internal_resource' for nil:NilClass
```

**Comment by no-reply** [comment]
this seems on the right track, in general. it might be simplified by using `shared_example` parameters as document here: https://relishapp.com/rspec/rspec-core/docs/example-groups/shared-examples

File: lib/hyrax/specs/shared_specs/indexers.rb [hunk 11]
```diff
@@ -79,9 +91,10 @@
   end
 
   describe '#to_solr' do
+    before { attributes.each { |k, v| the_resource.set_value(k, v) } }
     it 'indexes basic metadata' do
       expect(indexer.to_solr)
-        .to include(keyword_sim: a_collection_containing_exactly(*attributes[:keyword]),
+        .to include(keyword_sim:   a_collection_containing_exactly(*attributes[:keyword]),
```
Comment (elrayle) [inline lib/hyrax/specs/shared_specs/indexers.rb:97]: This change just makes the a_collection... align with a_collection... in the following 2 lines.  No real code change.

File: spec/support/book_resource.rb [hunk 12]
```diff
@@ -7,11 +7,11 @@ module Test
...
-      attribute :author,   Valkyrie::Types::String
-      attribute :created,  Valkyrie::Types::Date
-      attribute :isbn,     Valkyrie::Types::String
-      attribute :pubisher, Valkyrie::Types::String
-      attribute :title,    Valkyrie::Types::String
+      attribute :author,    Valkyrie::Types::String
+      attribute :created,   Valkyrie::Types::Date
+      attribute :isbn,      Valkyrie::Types::String
+      attribute :publisher, Valkyrie::Types::String
+      attribute :title,     Valkyrie::Types::String
```
Comment (elrayle) [inline spec/support/book_resource.rb:14]: Most of this is just an alignment change.  The only real change is fixing the typo from pubisher to publisher.

File: app/services/hyrax/thumbnail_path_service.rb [hunk 13]
```diff
@@ -5,7 +5,7 @@ class << self
       # @param [#id] object - to get the thumbnail for
       # @return [String] a path to the thumbnail
       def call(object)
-        return default_image unless object.try(:thumbnail_id)
+        return default_image unless object.try(:thumbnail_id)&.to_s&.present?
```
Comment (elrayle) [inline app/services/hyrax/thumbnail_path_service.rb:8]: For collections shared spec, the thumbnail_id was a Valkyrie ID with id="".  I don't know if this would happen in the application, but it doesn't hurt anything to have the extra code.  If thumbnail_id is nil, it will still use the default_image.

**Comment by elrayle** [comment]
@no-reply RE: parameters to shared specs...  I like the idea of making the parameters more transparent.  I don't think it would simplify the code if we still have the requirement of providing a default if the parameter does not have a value.  With a change to parameterized shared specs, would you see it as ok to remove the creation a default resource when one is not provided making parameters required?  That would simplify the code.

**Comment by elrayle** [comment]
Problem with parameters for shared specs.  You cannot use `let` or `before` to do setup on the parameters.  They have to be created on the `it_behaves_like` statement.

WORKS
```
  it_behaves_like 'a Hyrax::Resource indexer', described_class, Hyrax::Work.new
```

DOESN'T WORK  (error loading spec file)
```
  let(:indexer_class) { described_class }
  let(:resource) { Hyrax::Work.new }

  it_behaves_like 'a Hyrax::Resource indexer', indexer_class, resource
```

This example is simple.  But if you needed to make a modification to resource before sending it, there isn't a way to do that.

Also, you can't set up the resource once and have it used for all shared specs.  You have to copy the creation code to every it_behaves_like statement which isn't DRY.

```
  it_behaves_like 'a Hyrax::Resource indexer', described_class, Hyrax::Work.new
  it_behaves_like 'a Core metadata indexer', described_class, Hyrax::Work.new
  it_behaves_like 'a permission indexer', described_class, Hyrax::Work.new
  it_behaves_like 'a visibility indexer', described_class, Hyrax::Work.new
```

Any change to the resource you create requires every `it_behaves_like` to be updated.

**Comment by elrayle** [comment]
Example preprocessing from our app.  We default our works to public instead of private.  So I have to fake this to allow our indexer to pass the visibility indexer shared spec.

```
  context 'FAKE work defaulting to private visibility for shared specs' do
    before do
      Hyrax::VisibilityWriter.new(resource: resource)
        .assign_access_for(visibility: Hydra::AccessControls::AccessRight::VISIBILITY_TEXT_VALUE_PRIVATE)
    end
    it_behaves_like 'a visibility indexer'
  end
```

File: spec/indexers/hyrax/valkyrie_work_indexer_spec.rb [hunk 14]
```diff
@@ -21,4 +19,32 @@
...
+              end
+            end
+          end
+        end
+      end
+    end
+    after { Hyrax::Test.send(:remove_const, :Custom) }
+
+    let(:resource) { Hyrax.persister.save(resource: Hyrax::Test::Custom::Work.new(broader: ['term1', 'term2'])) }
+    let(:indexer_class) { Hyrax::Test::Custom::WorkIndexer }
```
Comment (jeremyf) [inline spec/indexers/hyrax/valkyrie_work_indexer_spec.rb:43]: Per conversation, this was already supported.

File: spec/indexers/hyrax/valkyrie_work_indexer_spec.rb [hunk 15]
```diff
@@ -21,4 +19,32 @@
...
+                solr_doc['broader_ssim'] = resource.broader.first
+              end
+            end
+          end
+        end
+      end
+    end
+    after { Hyrax::Test.send(:remove_const, :Custom) }
+
+    let(:resource) { Hyrax.persister.save(resource: Hyrax::Test::Custom::Work.new(broader: ['term1', 'term2'])) }
```
Comment (jeremyf) [inline spec/indexers/hyrax/valkyrie_work_indexer_spec.rb:42]: Per tech call, this is the line that the PR is supporting.

File: spec/indexers/hyrax/resource_indexer_spec.rb [hunk 16]
```diff
@@ -3,6 +3,7 @@
 require 'hyrax/specs/shared_specs'
 
 RSpec.describe Hyrax::ResourceIndexer do
+  let(:resource) { Hyrax::Resource.new }
```
Comment (jeremyf) [inline spec/indexers/hyrax/resource_indexer_spec.rb:6]: This is added to clarify and ensure the proper behavior. Prior to this, we had misleading successful test.

File: app/services/hyrax/thumbnail_path_service.rb (hunk 13 above)
Comment (jeremyf) [reply app/services/hyrax/thumbnail_path_service.rb:8]: This may be a separate commit. There may be other problems involved. @no-reply can you take action on this?

Comment (jeremyf) [reply app/services/hyrax/thumbnail_path_service.rb:8]: We may need to test that Valkyrie::ID with an empty string is not present.

File: lib/hyrax/specs/shared_specs/indexers.rb (hunk 7 above)
Comment (jeremyf) [reply lib/hyrax/specs/shared_specs/indexers.rb:5]: Based on conversations, the reason for these conditionals is for backwards compatibility.

Comment (jeremyf) [reply lib/hyrax/specs/shared_specs/indexers.rb:5]: As this was not released in a Hyrax version, we can remove the conditionals and tighten up the logic. Anyone that's using this as written will have a broken spec.

File: app/indexers/hyrax/valkyrie_collection_indexer.rb [hunk 17]
```diff
@@ -7,6 +7,7 @@ class ValkyrieCollectionIndexer < Hyrax::ValkyrieIndexer
     include Hyrax::ResourceIndexer
     include Hyrax::PermissionIndexer
     include Hyrax::VisibilityIndexer
+    include Hyrax::Indexer(:core_metadata)
```
Comment (elrayle) [inline app/indexers/hyrax/valkyrie_collection_indexer.rb:10]: Collection resource defines core metadata, so seems like the collection indexer should index it.

File: app/services/hyrax/thumbnail_path_service.rb [hunk 18]
```diff
@@ -5,7 +5,7 @@ class << self
       # @param [#id] object - to get the thumbnail for
       # @return [String] a path to the thumbnail
       def call(object)
-        return default_image unless object.try(:thumbnail_id)
+        return default_image unless object.try(:thumbnail_id)&.present?
```
Comment (elrayle) [inline app/services/hyrax/thumbnail_path_service.rb:8]: Catches the case where thumbnail_id is nil and when it is a Valkyrie::ID with an empty string.

File: lib/hyrax/specs/shared_specs/indexers.rb [hunk 19]
```diff
@@ -61,9 +82,37 @@
   end
 end
 
+RSpec.shared_examples 'a Core metadata indexer' do
```
Comment (elrayle) [inline lib/hyrax/specs/shared_specs/indexers.rb:85]: This is mostly the same with the similar changes for variable validation and removing the default resource.  It was moved to make the ordering of the various shared specs easier to follow.

File: spec/indexers/hyrax/resource_indexer_spec.rb (hunk 16 above)
Comment (elrayle) [reply spec/indexers/hyrax/resource_indexer_spec.rb:6]: resource is now required

File: lib/hyrax/specs/shared_specs/indexers.rb [hunk 20]
```diff
@@ -72,26 +121,40 @@
...
+    raise 'resource must be set with `let(:resource)` and is expected to be a kind of Hyrax::Work' unless defined?(resource) && resource.kind_of?(Hyrax::Work)
+    # optionally can pass in default_visibility by setting it with a let statement if your application changes the default; Hyrax defines this as 'restricted'
+    # See samvera/hyrda-head hydra-access-controls/app/models/concerns/hydra/access_controls/access_rights.rb for possible VISIBILITY_TEXT_VALUE_...'
+  end
+  subject(:indexer) { indexer_class.new(resource: resource) }
+
+  it_behaves_like 'a Hyrax::Resource indexer'
+  it_behaves_like 'a Core metadata indexer'
+  it_behaves_like 'a permission indexer'
+  it_behaves_like 'a visibility indexer'
```
Comment (elrayle) [inline lib/hyrax/specs/shared_specs/indexers.rb:146]: The shared spec defines the minimal requirements for a work instead of expecting the know which parts are required.

File: lib/hyrax/specs/shared_specs/indexers.rb [hunk 21]
```diff
@@ -72,26 +121,40 @@
...
 RSpec.shared_examples 'a Collection indexer' do
+  before do
+    raise 'indexer_class must be set with `let(:indexer_class)`' unless defined? indexer_class
+    raise 'resource must be set with `let(:resource)` and is expected to be a kind of Hyrax::PcdmCollection' unless defined?(resource) && resource.kind_of?(Hyrax::PcdmCollection)
+  end
   subject(:indexer) { indexer_class.new(resource: resource) }
-  let(:resource)    { Hyrax::PcdmCollection.new }
 
+  it_behaves_like 'a Hyrax::Resource indexer'
+  it_behaves_like 'a Core metadata indexer'
```
Comment (elrayle) [inline lib/hyrax/specs/shared_specs/indexers.rb:157]: The shared spec defines the minimal requirements for a collection instead of expecting the  individual collection tests to know which parts are required.

File: lib/hyrax/specs/shared_specs/indexers.rb [hunk 22]
```diff
@@ -107,21 +170,3 @@
     end
   end
 end
-
-RSpec.shared_examples 'a Core metadata indexer' do
```
Comment (elrayle) [inline lib/hyrax/specs/shared_specs/indexers.rb:111]: Moved higher in the file

File: spec/indexers/hyrax/valkyrie_collection_indexer_spec.rb [hunk 23]
```diff
@@ -3,9 +3,8 @@
 require 'hyrax/specs/shared_specs'
 
 RSpec.describe Hyrax::ValkyrieCollectionIndexer do
-  subject(:indexer) { described_class.new(resource: resource) }
+  let(:resource) { FactoryBot.valkyrie_create(:hyrax_collection) }
   let(:indexer_class) { described_class }
 
-  it_behaves_like 'a Hyrax::Resource indexer'
```
Comment (elrayle) [inline spec/indexers/hyrax/valkyrie_collection_indexer_spec.rb:9]: All shared specs representing the minimal required parts of a collection are now defined and tested in the shared spec 'a Collection indexer'.

File: spec/indexers/hyrax/valkyrie_work_indexer_spec.rb [hunk 24]
```diff
@@ -3,22 +3,48 @@
...
 RSpec.describe Hyrax::ValkyrieWorkIndexer do
-  let(:indexer) { described_class.new(resource: resource) }
   let(:resource) { FactoryBot.valkyrie_create(:hyrax_work) }
   let(:indexer_class) { described_class }
 
-  it_behaves_like 'a Hyrax::Resource indexer'
-  it_behaves_like 'a Core metadata indexer'
-  it_behaves_like 'a permission indexer'
-  it_behaves_like 'a visibility indexer'
+  it_behaves_like 'a Work indexer'
```
Comment (elrayle) [inline spec/indexers/hyrax/valkyrie_work_indexer_spec.rb:9]: All shared specs representing the minimal required parts of a work are now defined and tested in the shared spec 'a Work indexer'.

File: spec/hyrax/indexer_spec.rb [hunk 25]
```diff
@@ -13,6 +13,7 @@
   end
 
   context 'with core metadata schema' do
+    let(:resource) { work }
```
Comment (elrayle) [inline spec/hyrax/indexer_spec.rb:16]: resource is required for shared indexer specs

File: spec/indexers/hyrax/permission_indexer_spec.rb [hunk 26]
```diff
@@ -8,6 +8,7 @@
       include Hyrax::PermissionIndexer
     end
   end
+  let(:resource) { Hyrax::Resource.new }
```
Comment (elrayle) [inline spec/indexers/hyrax/permission_indexer_spec.rb:11]: resource is required for shared indexer specs

File: app/indexers/hyrax/valkyrie_collection_indexer.rb (hunk 17 above)
Comment (no-reply) [reply app/indexers/hyrax/valkyrie_collection_indexer.rb:10]: :+1:

File: app/services/hyrax/thumbnail_path_service.rb (hunk 18 above)
Comment (no-reply) [inline app/services/hyrax/thumbnail_path_service.rb:8]:
```suggestion
        return default_image unless object.try(:thumbnail_id).present?
```

Comment (no-reply) [reply app/services/hyrax/thumbnail_path_service.rb:8]: `nil` responds to `#present?`. i doubt this matters beyond visual clutter, but let's skip the `&`

File: lib/hyrax/specs/shared_specs/indexers.rb [hunk 27]
```diff
@@ -1,9 +1,19 @@
 # frozen_string_literal: true
 
+# These shared specs test various aspects of valkyrie based indexers by calling the #to_solr method.
```
Comment (no-reply) [inline lib/hyrax/specs/shared_specs/indexers.rb:3]: the docs are really appreciated!

File: lib/hyrax/specs/shared_specs/indexers.rb [hunk 28]
```diff
@@ -1,9 +1,19 @@
...
+#   * resource - a Hyrax::Resource that defines attributes and values consistent with the indexer
+#
+# NOTE: It is important that the resource has required values that the indexer #to_solr customizations expects to be available.
 RSpec.shared_examples 'a Hyrax::Resource indexer' do
-  subject(:indexer) { indexer_class.new(resource: resource) }
-  let(:ids)         { ['id1', 'id2'] }
-  let(:resource)    { Hyrax::Resource.new(alternate_ids: ids) }
+  before do
+    raise 'indexer_class must be set with `let(:indexer_class)`' unless defined? indexer_class
+    raise 'resource must be set with `let(:resource)` and is expected to be a kind of Hyrax::Resource' unless defined?(resource) && resource.kind_of?(Hyrax::Resource)
```
Comment (no-reply) [inline lib/hyrax/specs/shared_specs/indexers.rb:12]: i feel like the type checks might be heavy-handed, but let's keep them for now.
